### PR TITLE
Support passing `dict` as return value of functions

### PR DESCRIPTION
**Pull request by harshsingh-24**
Adds to #983 by giving full support of dictionary return type in LLVM and C backend respectively.

### CODE REVIEW

**Comment by harshsingh-24** [comment]
Hey @certik @Thirumalai-Shaktivel . Can you please help me out a little with this implementation? I am facing issues. If you can point me out a little to some portions.

**Comment by harshsingh-24** [comment]
@czgdp1807 I added `ASR::is_a<ASR::Dict_t>(*asr_type)` to the check `static inline bool is_llvm_struct(ASR::ttype_t* asr_type)` in `llvm_utils.h` to incorporate Dict return type using the above logic suggested by you.

**Comment by harshsingh-24** [comment]
Can you please check @Thirumalai-Shaktivel and @czgdp1807 as to why the build is failing on C Backend?

**Comment by czgdp1807** [comment]
> code generation error: Return type not supported in function 'fill_rollnumber2cpi'

You need to support `dict` as a return type in C backend as well (concerned files `asr_to_c.cpp`, `asr_to_c_cpp.h` and `asr_to_cpp.h`).

**Comment by harshsingh-24** [comment]
I did some changes and tried running - 
```
from lpython import i32

def fun() -> dict[i32, i32]:
    return {2: 4, 3: 6}

num : dict[i32, i32]
num = fun()
print(num[3])
```

The C backend throws error saying - 
```
(lp) root@Harsh:/mnt/d/B.Tech/workspace/wsl-workspace/lpython# ./src/bin/lpython examples/expr2.py --backend c
expr2__tmp__generated__.c: In function ‘_lpython_main_program’:
expr2__tmp__generated__.c:68:33: error: lvalue required as unary ‘&’ operand
   68 |     dict_deepcopy_dict_i32_i32_(&fun(), &num);
      |                                 ^
The command 'gcc -o expr2.out expr2__tmp__generated__.c  -I /mnt/d/B.Tech/workspace/wsl-workspace/lpython/src/bin/../libasr/runtime' failed.
```

The generated code sub-section as shown below for dict is - 
```
void _lpython_main_program()
{
    dict_deepcopy_dict_i32_i32_(&fun(), &num);
    printf("%d\n", dict_get_item_dict_i32_i32_(&num, 3));
}
```

However, it should be similar to lists - 
```
void _lpython_main_program()
{
    struct list_i32 constname01 = fun();
    list_deepcopy_i32(&constname01, &num);
    printf("%d\n", num.data[2]);
}
```

**Comment by harshsingh-24** [comment]
Can you please review ? @Thirumalai-Shaktivel @czgdp1807 @certik

File: integration_tests/test_dict_02.py [hunk 1]
```diff
@@ -27,7 +35,6 @@ def test_dict():
     for i in range(1000, 1000 + size, 7):
         assert abs(rollnumber2cpi[i] + f64(i)/100.0 + 5.0) <= 1e-12
 
-
     assert abs(rollnumber2cpi[0] - 1.1) <= 1e-12
 
 test_dict()
```
Comment (certik) [inline integration_tests/test_dict_02.py:40]: I would add new tests rather than modifying the existing ones, as I think we want to be testing both.

Comment (harshsingh-24) [reply integration_tests/test_dict_02.py]: Done with the changes. Reverted the changes in original test cases. Made new test cases and added them to `CmakeLists`. @certik

**Comment by harshsingh-24** [comment]
Can you review? @czgdp1807 @certik . CI tests have passed.

File: integration_tests/test_dict_06.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    
+    rollnumber2cpi[0] = 1.1
+    for i in range(1000, 1000 + size):
+        rollnumber2cpi[i] = float(i/100.0 + 5.0)
+
+    return rollnumber2cpi
+
+def test_dict():
+    i: i32
+    size: i32 = 1000
```
Comment (czgdp1807) [inline integration_tests/test_dict_06.py:15]:
```suggestion
    size: i32 = 200
```

Smaller value for quickly running the tests in CPython. `1000` is already used in other test of `dict` so they will ensure proper testing already.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

File: integration_tests/test_dict_06.py (hunk 2 above)
Comment (harshsingh-24) [reply integration_tests/test_dict_06.py]: I did the changes.

Comment (harshsingh-24) [reply integration_tests/test_dict_06.py]: @czgdp1807 @certik review.

**Review by certik — APPROVED** [review]
This looks good to me, but I would like @czgdp1807 to also review. The history needs polishing or squashing. I rebased on top of the latest master.

**Review by czgdp1807 — APPROVED** [review]

**Pull request merged**